Panic in PatchMake

Hitting a slice bounds out of range panic

````
panic: runtime error: slice bounds out of range

goroutine 47882 [running]:
panic(0xea1500, 0xc820034020)
	/usr/local/go1.6.3.src/src/runtime/panic.go:481 +0x3e6
..gojsondiff/vendor/github.com/sergi/go-diff/diffmatchpatch.(*DiffMatchPatch).patchMake2(0xc8215e5c30, 0xc820f708c0, 0x1e, 0xc821594240, 0xa, 0x18, 0x0, 0x0, 0x0)
	..gojsondiff/vendor/github.com/sergi/go-diff/diffmatchpatch/dmp.go:1810 +0x647
github.com/apcera/cntm-deps/gojsondiff/vendor/github.com/sergi/go-diff/diffmatchpatch.(*DiffMatchPatch).PatchMake(0xc8215e5c30, 0xc8215e5bb8, 0x2, 0x2, 0x0, 0x0, 0x0)
	..gojsondiff/vendor/github.com/sergi/go-diff/diffmatchpatch/dmp.go:1768 +0x486
.
.
````
 Able to reproduce this by feeding ```(left=2016-09-01T03:07:14.807830741Z, right=2016-09-01T03:07:15.154800781Z)``` to ```PatchMake```.

Following diff being applied to left string:``` [{0 2016-09-01T03:07:1} {1 5.15} *{0 .154} {-1 .} {0 80} {1 0} {0 78} {-1 3074} {0 1Z}]```

while the diff being applied should have been ``` [{0 2016-09-01T03:07:1} {1 5.15} *{0 4} {-1 .} {0 80} {1 0} {0 78} {-1 3074} {0 1Z}]```

Thank you for your time.

--- a/diffmatchpatch/dmp.go
+++ b/diffmatchpatch/dmp.go
@@ -967,7 +967,7 @@ func (dmp *DiffMatchPatch) DiffCleanupSemantic(diffs []Diff) []Diff {
 					float64(overlapLength2) >= float64(len(insertion))/2 {
 					// Reverse overlap found.
 					// Insert an equality and swap and trim the surrounding edits.
-					overlap := Diff{DiffEqual, insertion[overlapLength2:]}
+					overlap := Diff{DiffEqual, insertion[len(insertion)-overlapLength2:]}
 					diffs = append(
 						diffs[:pointer],
 						append([]Diff{overlap}, diffs[pointer:]...)...)

--- a/diffmatchpatch/dmp_test.go
+++ b/diffmatchpatch/dmp_test.go
@@ -563,6 +563,29 @@ func Test_diffCleanupSemantic(t *testing.T) {
 		Diff{DiffEqual, "1234"},
 		Diff{DiffDelete, "wxyz"}}, diffs)
 
+	// No elimination #3.
+	diffs = []Diff{
+		Diff{DiffEqual, "2016-09-01T03:07:1"},
+		Diff{DiffInsert, "5.15"},
+		Diff{DiffEqual, "4"},
+		Diff{DiffDelete, "."},
+		Diff{DiffEqual, "80"},
+		Diff{DiffInsert, "0"},
+		Diff{DiffEqual, "78"},
+		Diff{DiffDelete, "3074"},
+		Diff{DiffEqual, "1Z"}}
+	diffs = dmp.DiffCleanupSemantic(diffs)
+	assertDiffEqual(t, []Diff{
+		Diff{DiffEqual, "2016-09-01T03:07:1"},
+		Diff{DiffInsert, "5.15"},
+		Diff{DiffEqual, "4"},
+		Diff{DiffDelete, "."},
+		Diff{DiffEqual, "80"},
+		Diff{DiffInsert, "0"},
+		Diff{DiffEqual, "78"},
+		Diff{DiffDelete, "3074"},
+		Diff{DiffEqual, "1Z"}}, diffs)
+
 	// Simple elimination.
 	diffs = []Diff{
 		Diff{DiffDelete, "a"},

--- a/diffmatchpatch/dmp_test.go
+++ b/diffmatchpatch/dmp_test.go
@@ -1268,6 +1268,9 @@ func Test_patch_make(t *testing.T) {
 	expectedPatch = "@@ -573,28 +573,31 @@\n cdefabcdefabcdefabcdefabcdef\n+123\n"
 	patches = dmp.PatchMake(text1, text2)
 	assert.Equal(t, expectedPatch, dmp.PatchToText(patches), "patch_make: Long string with repeats.")
+
+	patches = dmp.PatchMake("2016-09-01T03:07:14.807830741Z", "2016-09-01T03:07:15.154800781Z")
+	assert.Equal(t, "@@ -15,16 +15,16 @@\n 07:1\n+5.15\n 4\n-.\n 80\n+0\n 78\n-3074\n 1Z\n", dmp.PatchToText(patches), "patch_make: Corner case of #31 fixed by #32")
 }
 
 func Test_PatchSplitMax(t *testing.T) {

